crate/server_database/src/stores/redis/permissions.rs

```diff
@@ -66,266 +105,180 @@ impl Triple {
 
-    fn try_from(value: &Location) -> Result<Self, Self::Error> {
-        let value = String::from_utf8((value).to_vec())?;
-        let mut parts = value.split("::");
-        let uid = parts.next().ok_or_else(|| {
-            DbError::ConversionError(format!("invalid permissions triple: {parts:?}"))
-        })?;
-        let user_id = parts.next().ok_or_else(|| {
-            DbError::ConversionError(format!("invalid permissions triple: {parts:?}"))
-        })?;
-        let permission = parts.next().ok_or_else(|| {
-            DbError::ConversionError(format!("invalid permissions triple: {parts:?}"))
+    fn length(&self) -> usize {
+        // obj_uid + user_id + permission
+        self.obj_uid.0.length() + self.user_id.0.length() + 1 // 1 byte is KmipOperation's enum size
+    }
+
+    fn write(
+        &self,
+        ser: &mut cosmian_kms_crypto::reexport::cosmian_crypto_core::bytes_ser_de::Serializer, // full dependency path spec if needed to avoid collisions
+    ) -> Result<usize, Self::Error> {
+        let mut written = 0;
+        written += ser.write(&self.obj_uid.0)?;
+        written += ser.write(&self.user_id.0)?;
+        written += ser.write_vec(&[self.permission as u8])?;
+        Ok(written)
+    }
+
+    fn read(
+        de: &mut cosmian_kms_crypto::reexport::cosmian_crypto_core::bytes_ser_de::Deserializer,
+    ) -> Result<Self, Self::Error> {
+        let obj_uid = ObjectUid(String::from_utf8(de.read_vec()?)?);
+        let user_id = UserId(String::from_utf8(de.read_vec()?)?);
+        let perm_vec = de.read_vec()?;
+        let permission = KmipOperation::from_repr(perm_vec[0]).ok_or_else(|| {
+            DbError::ConversionError(format!("Invalid KmipOperation value: {}", perm_vec[0]))
         })?;
-        Ok(Self {
-            obj_uid: uid.to_owned(),
-            user_id: user_id.to_owned(),
-            permission: serde_json::from_str(permission)?,
+        Ok(PermTriple {
+            obj_uid,
+            user_id,
+            permission,
         })
     }
 }
 
-impl TryFrom<&Triple> for Location {
+impl TryFrom<&IndexedValue> for PermTriple {
     type Error = DbError;
 
-    fn try_from(value: &Triple) -> Result<Self, Self::Error> {
-        Ok(Self::from(
-            format!(
-                "{}::{}::{}",
-                value.obj_uid,
-                value.user_id,
-                serde_json::to_string(&value.permission)?
-            )
-            .into_bytes(),
+    fn try_from(value: &IndexedValue) -> Result<Self, Self::Error> {
+        cosmian_kms_crypto::reexport::cosmian_crypto_core::bytes_ser_de::Serializable::deserialize(
+            value,
+        )
+    }
+}
+
+impl TryFrom<&PermTriple> for IndexedValue {
+    type Error = DbError;
+
+    fn try_from(value: &PermTriple) -> Result<Self, Self::Error> {
+        Ok(IndexedValue::from(
+          cosmian_kms_crypto::reexport::cosmian_crypto_core::bytes_ser_de::Serializable::serialize(
+               value,
+           )?.to_vec()
```

Comment:
Note for future work: we need to implement an unsafe serializer in `cosmian_crypto_core` to avoid this cloning (it appears everywhere in the KMS).